nav2_costmap_2d/src/costmap_cleaner.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <memory>
+
+#include "nav2_costmap_2d/costmap_cleaner.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+namespace nav2_costmap_2d
+{
+
+using std::vector;
+using std::string;
+using std::shared_ptr;
+using std::any_of;
+using CleanCostmap = nav2_msgs::srv::CleanCostmap;
+
+CostmapCleaner::CostmapCleaner(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
+: node_(node), costmap_(costmap)
+{
+  node_->get_parameter_or_set("cleanable_layers", cleanable_layers_, {"obstacle_layer"});
+
+  server_ = node_->create_service<CleanCostmap>("clean_" + costmap->getName(),
+      std::bind(&CostmapCleaner::cleanCallback, this,
+      std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+}
+
+void CostmapCleaner::cleanCallback(
+  const shared_ptr<rmw_request_id_t>/*request_header*/,
+  const shared_ptr<CleanCostmap::Request> request,
+  const shared_ptr<CleanCostmap::Response>/*response*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to clean " + costmap_->getName());
+
+  if (costmap_ == nullptr) {
+    RCLCPP_ERROR(node_->get_logger(), "Costmap is undefined. Doing nothing.");
+    return;
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "Will proceed with cleaning the costmap");
+
+  clean(request->reset_distance);
+}
+
+void CostmapCleaner::clean(const double reset_distance)
+{
+  double x, y;
+
+  if (!getPose(x, y)) {
+    RCLCPP_ERROR(node_->get_logger(), "Cannot clean map because robot pose cannot be retrieved.");
+    return;
+  }
+
+  auto layers = costmap_->getLayeredCostmap()->getPlugins();
+
+  for (auto & layer : *layers) {
+    auto name = getLayerName(*layer);
+
+    if (any_of(
```

Comment:
Suggestion: this is a more readability thing, if this was implemented using usual C++98 methods, this would be far easier for the typical person to understand, this takes me about 2-3 passes to understand the full intention of a pretty simple operation.

Reply:
I updated the code to improve the readability of this line, while still using `std::any_of`.